tuf/encoding/asn1_metadata_definitions.py

```diff
@@ -0,0 +1,205 @@
+import pyasn1.type.namedtype as namedtype
+# Likely imports for later:
+# import pyasn1.type.namedval as namedval
+# import pyasn1.type.tag as tag
+from pyasn1.type.namedtype import NamedType, NamedTypes, DefaultedNamedType
+
+# Maximum integer value when bounding integer values below.
+# It does not seem possible to place a minimum without placing a maximum.
+# For thresholds, versions, timestamps, etc., we want only non-negative
+# integers, so we're stuck setting a maximum (or not placing any constraints
+# at all in the ASN.1 metadata definition itself, would would also be fine).
+MAX = 2**32-1
+
+
+## Common types, for use in the various metadata types
+
+# String types
+class SignatureMethod(char.VisibleString): pass
+class RoleName(char.VisibleString): pass
+class Filename(char.VisibleString): pass
+class HashFunction(char.VisibleString): pass
+class KeyType(char.VisibleString): pass
+
+
+# Binary data types
+class KeyID(univ.OctetString): pass
+class Path(char.VisibleString): pass
+
+
+# Integer types
+class Length(univ.Integer):
+  subtypeSpec = constraint.ValueRangeConstraint(0, MAX)
+
+class Threshold(univ.Integer):
+  subtypeSpec = constraint.ValueRangeConstraint(0, MAX)
+
+class Version(univ.Integer):
+    subtypeSpec = constraint.ValueRangeConstraint(0, MAX)
+
+class UTCDateTime(univ.Integer):
+  subtypeSpec = constraint.ValueRangeConstraint(0, MAX)
+
+
+# More complex common types follow.
+class Signature(univ.Sequence):
+  componentType = NamedTypes(
+      NamedType('keyid', KeyID()),
+      NamedType('method', SignatureMethod()),
+      NamedType('value', univ.OctetString()))
+
+class Hash(univ.Sequence):
+  componentType = NamedTypes(
+      NamedType('function', HashFunction()),
+      NamedType('digest', univ.OctetString()))
+
+class PublicKey(univ.Sequence):
+  componentType = NamedTypes(
+      NamedType('publicKeyID', KeyID()),
+      NamedType('publicKeyType', KeyType()),
+      NamedType('publicKeyValue', univ.OctetString()))
+
+
+
+
+## Types used only in Root metadata
+class RootMetadata(univ.Sequence):
+  componentType = NamedTypes(
+      NamedType('type', RoleName()),
+      NamedType('expires', UTCDateTime()),
+      NamedType('version', Version()),
+      NamedType('consistent-snapshot', univ.Boolean()),
+      NamedType('num-keys', Length()),
+      NamedType('keys', univ.SequenceOf(componentType=PublicKey())),
+      NamedType('num-roles', Length()),
+      NamedType('roles', univ.SequenceOf(componentType=TopLevelDelegation())))
+
+class TopLevelDelegation(univ.Sequence):
+  componentType = NamedTypes(
+      NamedType('role', RoleName()),
+      NamedType('num-keyids', Length()),
+      NamedType('keyids', univ.SequenceOf(componentType=KeyID())),
+      NamedType('threshold', Threshold()))
+
+
+
+
+## Types used only in Timestamp metadata
+class HashOfSnapshot(univ.Sequence):
+  componentType = NamedTypes(
+      NamedType('filename', FileName()),
+      NamedType('num-hashes', Length()),
```

Comment:
Is there a reason Length() is reused for num-keys, num-keyids, and num-hashes? To clarify, is it possible to have a more descriptive name?

Reply:
I guess we could call it SequenceLength. Do you think that's better? Or do you think it's clearer if I just exclude it and use an integer primitive directly in each of these places?